contributors/design-proposals/cpu-affinity-numa-awareness.md

```diff
@@ -0,0 +1,216 @@
+	// the scheduler simply schedules this pod onto that node, assuming that it fits resource
+	// requirements.
+	// +optional
+	NodeName string
+	// Identifies a NUMA node that affines the pod.  If it is non-empty, the value must
+	// correspond to a particular NUMA node on the same node that the pod is scheduled against.
+	// This value is only set if either the `CPUAffinity` or `NUMACPUAffinity` tolerations
+	// are present on the pod.
+	// +optional
+	NUMANodeID string
+	// CPUAffinity controls the CPU affinity of the executed pod.
+	// If it is non-empty, the value must correspond to a particular set
+	// of CPU cores in the matching NUMA node on the machine that the pod is scheduled against.
+	// This value is only set if either the `CPUAffinity` or `NUMACPUAffinity` tolerations
+	// are present on the pod.
+	// The values are expressed in the List Format syntax specified here:
+	// here: http://man7.org/linux/man-pages/man7/cpuset.7.html
+	// +optional
+	CPUAffinity string	
+```
+
+### REST API changes
+
+The `/pod/<pod-name>/bind` operation will allow updating the NUMA and CPU
+affinity values.  The same permissions required to schedule a pod to a
+node in the cluster will be required to bind a pod to a particular NUMA node
+and CPU set.
+
+### Tolerations
+
+Pods that require CPU and NUMA affinity prior to execution must set the
+appropriate `Tolerations` for the associated taints.
+
+### Multiple containers
+
+If a pod has multiple containers, the set of containers must all fit
+a specific NUMA node, and the set of affined CPUs are shared among containers.
+
+Pod level cgroups are used to actually affine the container to the specified
+CPU set.
+
+## Resource Quota changes
+
+Operators must be able to limit the consumption of dedicated CPU cores via quota.
+
+## Kubelet changes
+
+The `kubelet` will enforce the presence of the required pod tolerations assigned to the node.
+
+The `kubelet` will pend the execution of any pod that is assigned to the node, but has
+not populated the required fields for a particular toleration.
+
+* If the toleration `CPUAffinity` is present on a `Pod`, the pod will not start
+any associated container until the `Pod.Spec.CPUAffinity` is populated.
+* If the toleration `NUMAAffinity` is present on a `Pod`, the pod will not start
+any associated container until the `Pod.Spec.NUMANodeID` is populated.
+
+The delayed execution of the pod enables both a single and dual-phase scheduler to
+place pods on a particular NUMA node and set of CPU cores.
+
+## Required work:
+
+1. pod level cgroup support roll-out
+1. implement support for `NoScheduleNoAdmitNoExecute` taint effect
+1. expose NUMA topology in cAdvisor
+1. expose NUMA topology in node status
+1. pod level cgroup support for enabling cpu set
+
+## Future considerations
+
+1. Author `NUMATopologyPredicate` in scheduler to enable NUMA aware scheduling.
+1. Restrict vertical autoscaling of CPU and NUMA affined workloads.
```

Comment:
Does this have to do with limiting dynamic logical core assignments, or specifically fixing the number of cores allocated for the lifetime of a pod? If the latter, what's the rationale for it? The default vertical auto-scaler (non-numa-aware) could emit impossible allocations (spanning sockets, fractional cores, etc) but you could also imagine a numa-aware vertical auto-scaler that can do the right thing.